### PR TITLE
[C++ API] Add backward() to Tensor and Variable 

### PR DESCRIPTION
**Pull request by goldsborough**
A sort of small and uncomplicated change, but nice for the C++ world: adds `backward()` to the "autograd API" of `at::Tensor` and implements it for `VariableImpl` to execute the default autograd engine on the variable.

Main code is in `autograd/variable.cpp`.

I've removed the equivalent free function from autogradpp.

@zdevito @colesbury @apaszke @ebetica

### CODE REVIEW

File: torch/csrc/autograd/variable.cpp [hunk 1]
```diff
@@ -109,6 +111,22 @@ void Variable::Impl::detach_() {
...
+  std::vector<Edge> edges;
+  edges.emplace_back(grad_fn_, output_nr_);
+
+  std::vector<Variable> inputs;
+  if (!gradient.has_value()) {
+    gradient = make_variable(at::ones_like(data_), /*requires_grad=*/false);
+  }
+  inputs.push_back(std::move(as_variable_ref(*gradient)));
+
+  Engine::get_default_engine().execute(edges, inputs, keep_graph, create_graph);
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: nit: would be simpler to just do `{gradient}` instead of constructing `inputs` above

File: torch/csrc/autograd/variable.h [hunk 2]
```diff
@@ -22,15 +22,18 @@ namespace torch {
...
-namespace jit { namespace tracer {
+namespace jit {
+namespace tracer {
 // Has to be forward declared because tracer_state.h has a dependency on
 // variable.h.
 struct ValueTracingStateElem;
 using ValueTracingState = std::list<ValueTracingStateElem>;
-}} // namespace jit::tracer
+} // namespace tracer
+} // namespace jit
```
Comment (apaszke) [inline torch/csrc/autograd/variable.h]: Come on man. Those changes are really unnecessary, and they're not even consistent with what we usually do. Why are they part of this diff?!

**Review by apaszke — APPROVED** [review]
LGTM

Splitting namespaces into multiple lines is really unnecessary. Such changes don't add any value to our code, and really make it less consistent. Please don't include such changes in the future, unless we establish a code format policy, and will do a full codebase refactor.

File: torch/csrc/autograd/variable.cpp [hunk 3]
```diff
@@ -109,6 +111,22 @@ void Variable::Impl::detach_() {
   output_nr_ = 0;
 }
 
+void Variable::Impl::backward(
```
Comment (ebetica) [inline torch/csrc/autograd/variable.cpp]: nit: Can we give the C++ API the same argument set as this one?

Comment (goldsborough) [reply torch/csrc/autograd/variable.cpp]: What do you mean? I've removed the C++ API one, so you'd only use this method

File: torch/csrc/autograd/variable.h (hunk 2 above)
Comment (goldsborough) [reply torch/csrc/autograd/variable.h]: The `.clang-format` got changed recently to un-compact namespaces. Since I rewrote this file I had it clang-formatted entirely at the time, and assumed it safe to re-format it. But I agree this goes against autograd style. I will fix it

File: torch/csrc/autograd/variable.cpp (hunk 3 above)
Comment (ebetica) [reply torch/csrc/autograd/variable.cpp]: Ah my bad, I missed it. LGTM

**Comment by ezyang** [comment]
FYI, according to GitHub this was merged without CI going green. Even after the subsequent fix, Windows is still failing (https://github.com/pytorch/pytorch/pull/7762).

**Comment by goldsborough** [comment]
Yeah I pushed a quick two line formatting fix after everything was already green, and merged it immediately. The formatting fix had the missing braces
